### PR TITLE
flash: spi_nor: don't hard loop in `wait_until_ready`

### PR DESCRIPTION
**Pull request by JordanYates**
Don't monopolise the CPU in `spi_nor_wait_until_ready`. For slow flash
chips, operations can take minutes (Full chip erase on MX25R is listed
as 120s typical, 240s max).

### CODE REVIEW

**Comment by Laczen** [comment]
@JordanYates Nice addition. Should this be wrapped into some config to allow excluding the k_sleep in cases where it is allowed to block the cpu but code size is more important (e.g. a bootloader) ?

**Comment by JordanYates** [comment]
> @JordanYates Nice addition. Should this be wrapped into some config to allow excluding the k_sleep in cases where it is allowed to block the cpu but code size is more important (e.g. a bootloader) ?

I can't imagine that the extra code amounts to more than a few dozen bytes, it doesn't seem worth it in my opinion.

**Comment by Laczen** [comment]
> > @JordanYates Nice addition. Should this be wrapped into some config to allow excluding the k_sleep in cases where it is allowed to block the cpu but code size is more important (e.g. a bootloader) ?
> 
> I can't imagine that the extra code amounts to more than a few dozen bytes, it doesn't seem worth it in my opinion.

Adding a `k_sleep()` function adds about 448 byte on a nrf52dk_nrf52832, and if this requires SYS_TIMER to be enabled this adds another 3682 byte.

File: drivers/flash/spi_nor.c [hunk 1]
```diff
@@ -186,6 +186,13 @@ static const struct jesd216_erase_type minimal_erase_types[JESD216_NUM_ERASE_TYP
 };
 #endif /* CONFIG_SPI_NOR_SFDP_MINIMAL */
 
+/* Different operations should poll the WIP bit at different rates */
+enum wait_ready_op_type {
+	WAIT_READY_REGISTER     = 0,
+	WAIT_READY_WRITE        = 1,
+	WAIT_READY_ERASE        = 2,
+};
```
Comment (de-nordic) [inline drivers/flash/spi_nor.c:194]: I was already thinking on how to address the issue you are addressing here, with addition to hardware timeouts and some other stuff, but there are currently some Zephyr design hurdles I can not overcome easily, and I will say no here.
Just parametrize the `spi_nor_wait_until_ready` to take the timeout directly.
You can still use the above but define them directly as timeouts, i.e. `#define WAIT_READY_REGISTER K_NO_WAIT`.

Comment (JordanYates) [reply drivers/flash/spi_nor.c]: Fair enough, the enum would allow for different backoff schemes, but thats not a high priority

**Comment by JordanYates** [comment]
> Adding a `k_sleep()` function adds about 448 byte on a nrf52dk_nrf52832, and if this requires SYS_TIMER to be enabled this adds another 3682 byte.

Is this likely to be the only `k_sleep` in an application/bootloader? Happy to take suggestions on how you would control the inclusion. A kconfig defaulted by `BOOTLOADER`?

**Comment by de-nordic** [comment]
> > Adding a `k_sleep()` function adds about 448 byte on a nrf52dk_nrf52832, and if this requires SYS_TIMER to be enabled this adds another 3682 byte.
> 
> Is this likely to be the only `k_sleep` in an application/bootloader? Happy to take suggestions on how you would control the inclusion. A kconfig defaulted by `BOOTLOADER`?

Yeah, we have to be able to control this to cut down used flash on smaller devices. This is kinda tradeoff.
Bootloader would rather have not that, as it does not care how long it has to wait.

**Comment by Laczen** [comment]
> > Adding a `k_sleep()` function adds about 448 byte on a nrf52dk_nrf52832, and if this requires SYS_TIMER to be enabled this adds another 3682 byte.
> 
> Is this likely to be the only `k_sleep` in an application/bootloader? Happy to take suggestions on how you would control the inclusion. A kconfig defaulted by `BOOTLOADER`?

proposal: `CONFIG_SPI_NOR_SLEEP_WHILE_WAITING_UNTIL_READY` default to y

**Review by Laczen — APPROVED** [review]
Thanks for the proposal and the changes, this is a very nice addition.

**Review by de-nordic — APPROVED** [review]
LGTM, thanks.

@JordanYates Would it be possible for you to create another PR addressing the addition of Kconfig, and the change, in release notes?

**Review by fabiobaltieri — APPROVED** [review]

**Comment by JordanYates** [comment]
> LGTM, thanks.
> 
> @JordanYates Would it be possible for you to create another PR addressing the addition of Kconfig, and the change, in release notes?

Done in #68086

File: drivers/flash/spi_nor.c [hunk 2]
```diff
@@ -395,17 +402,27 @@ static int spi_nor_access(const struct device *const dev,
...
-static int spi_nor_wait_until_ready(const struct device *dev)
+static int spi_nor_wait_until_ready(const struct device *dev, k_timeout_t poll_delay)
 {
 	int ret;
 	uint8_t reg;
 
-	do {
-		ret = spi_nor_cmd_read(dev, SPI_NOR_CMD_RDSR, &reg, sizeof(reg));
-	} while (!ret && (reg & SPI_NOR_WIP_BIT));
+	ARG_UNUSED(poll_delay);
```
Comment (GeorgeCGV) [inline drivers/flash/spi_nor.c:413]: Missing `#ifndef CONFIG_SPI_NOR_SLEEP_WHILE_WAITING_UNTIL_READY`

**Pull request merged**